### ISSUE TITLE
[cuperino_http] Crash in cupertino_http

### ISSUE THREAD

**Issue by knaeckeKami** [dart-lang/http#887]
I recently rolled out cupertino_http for a small percentage of users in a production app and I see some crashes associated with it.

I did not reproduce it myself yet, but I thought I'd share the stack trace here:

```
OS Version: iOS 15.7.3 (19H307)
Report Version: 104

Exception Type: EXC_BAD_ACCESS (SIGBUS)
Exception Codes: BUS_NOOP at 0x0000000d531fc190
Crashed Thread: 15

Application Specific Information:
lock > ryView:withSize: >
Attempted to dereference garbage pointer 0xd531fc190.

Thread 15 Crashed:
0   libobjc.A.dylib                 0x32fa99cf4         objc_msgSend
1   cupertino_http                  0x102986028         -[CUPHTTPClientDelegate URLSession:dataTask:didReceiveResponse:completionHandler:] (CUPHTTPClientDelegate.m:126)
2   CFNetwork                       0x30197d4ec         CFURLConnectionCreateWithProperties
3   Foundation                      0x3037c8398         __NSBLOCKOPERATION_IS_CALLING_OUT_TO_A_BLOCK__
4   Foundation                      0x3037d9a6c         -[NSBlockOperation main]
5   Foundation                      0x3037b3a88         __NSOPERATION_IS_INVOKING_MAIN__
6   Foundation                      0x3037c3fc8         -[NSOperation start]
7   Foundation                      0x3037c7494         __NSOPERATIONQUEUE_IS_STARTING_AN_OPERATION__
8   Foundation                      0x3037d4cc8         __NSOQSchedule_f
9   libdispatch.dylib               0x300341540         _dispatch_block_async_invoke2
10  libdispatch.dylib               0x300394090         _dispatch_client_callout
11  libdispatch.dylib               0x300336bb4         _dispatch_continuation_pop$VARIANT$mp
12  libdispatch.dylib               0x3003362dc         _dispatch_async_redirect_invoke
13  libdispatch.dylib               0x300343b90         _dispatch_root_queue_drain
14  libdispatch.dylib               0x300344398         _dispatch_worker_thread2
15  libsystem_pthread.dylib         0x3b712fdd0         _pthread_wqthread

```

Additional info: I use cupertino_http in a separate isolate, not in the Main isolate. Not sure if related.

**Comment by brianquinlan** [comment]
Hey @knaeckeKami, thanks for the great bug report. I'm looking into it now.

**Comment by brianquinlan** [comment]
I **think** that I screwed up the reference count management. Let me try to verify.

**Comment by knaeckeKami** [comment]
Thanks your the prompt fix!

**Comment by brianquinlan** [comment]
@knaeckeKami  I'm going to do a release in a few minutes, would you mind testing it?

**Comment by knaeckeKami** [comment]
will do!

**Comment by knaeckeKami** [comment]
We have enabled native http for all users for a 2 weeks now and have not seen any crashes related to that, seems good

**Comment by brianquinlan** [comment]
@knaeckeKami Awesome! If you don't mind me asking, in what app are you using cupertino_http?

**Comment by knaeckeKami** [comment]
We use it for Hello Inside, an app that aims to offer you insights into how your body reacts to certain foods, habits, etc. using continuous glucose monitoring. 

https://helloinside.com/hello-hormones/

The reason why we tried cupertino_http is that a small fraction of our users reported, that the app behaves as if they were offline all time time.
The native parts of the app (sentry error reporting, customer support) worked though, we saw that every HTTP call was terminated with "Connection Reset By Peer". 

It was probably an issue with proxies, metered connections, or even anti-malware systems blocking the calls (the users are not particularly tech-savvy, so they could not really tell us more about their network).

But since native HTTP calls using Sentry etc. worked, we decided the enable cupertino_http for them to see if this solved their issue, and it did.
